replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,335 @@
+    this.local = local;
+    this.siteManager = siteManager;
+    this.nodeAdapters = nodeAdapters;
+  }
+
+  /**
+   * Begin monitoring a queue for tasks. The handler of this worker is responsible for halting this
+   * thread by calling interrupt on this worker's thread, finishing the transfer of the current task
+   * if necessary.
+   */
+  @Override
+  public void run() {
+    while (running) {
+      Task task = null;
+      try {
+        task = queue.take();
+        this.processing = true;
+        if (supported(task)) {
+          this.handle(task);
+        } else {
+          task.unlock();
+        }
+        this.processing = false;
+      } catch (InterruptedException e) {
+        if (task != null) {
+          try {
+            task.unlock();
+          } catch (InterruptedException ignore) { // fallthrough
+          }
+        }
+        interrupt();
+        break;
+      } finally {
+        try {
+          local.close();
+        } catch (IOException e) {
+          LOGGER.debug("Failed to close adapter", e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Set this workers running flag to false if it is not currently transferring an item
+   *
+   * @return {@code true} if the worker was cancelled, otherwise {@code false}
+   */
+  boolean cancelIfProcessing() {
+    if (processing) {
+      this.running = false;
+      return true;
+    }
+    return false;
+  }
+
+  private boolean supported(Task task) {
+    if (!OperationType.UNKNOWN.equals(task.getOperation())) {
+      return false;
+    }
+
+    List<MetadataInfo> infos = task.metadatas().collect(Collectors.toList());
+    if (infos.size() > 1) {
+      LOGGER.warn("More than one metadata is currently not supported by this worker");
+      return false;
+    }
+
+    DDFMetadataInfo ddfInfo = infos.stream().filter(DDFMetadataInfo.class::isInstance)
+        .map(DDFMetadataInfo.class::cast)
+        .findFirst()
+        .orElse(null);
+
+    if (ddfInfo == null) {
+      LOGGER.info("Unable to handle any of the metadatas received in this task");
+      return false;
+    }
+
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    NodeAdapter destination = getDestinationAdapter(task);
+    if (destination == null) {
+      task.unlock();
+      return;
+    }
+
+    if (!adaptersAvailable(task, destination)) {
+      return;
+    }
+
+    List<DDFMetadataInfo> infos = task.metadatas()
+        .filter(DDFMetadataInfo.class::isInstance)
+        .map(DDFMetadataInfo.class::cast)
+        .collect(Collectors.toList());
+
+    final DDFMetadataInfo info = infos.get(0);
+    final Metadata metadata = new MetadataImpl(((Map) info.getData()).get("raw-metadata"),
+        info.getDataClass(), task.getId(),
+        Date.from(info.getLastModified()));
+
+    final OperationType type = task.getOperation();
+    switch (type) {
+      case HARVEST:
+      case REPLICATE_FROM:
+        doCreate(task, metadata, destination, local);
```

Comment:
If the query service still works how I think it is then we will also get updates via the queue as well. I assume we are expecting the create to fail if we have already created it (based on correlation id). How are we going to handle updates that come through and just keep failing? Might not be a concern for the worker but we need to make sure we handle that. Ion needs to probably handle updates to solve this.

Reply:
AFAIK updates will not be supported in this iteration. The query service does not have an appropriate operation type to use.

Reply:
updates should be treated down the road as upserts. This is purposely why we don't differentiate updates from creates in the type of operations. For DDF, it would be up to the worker to handle that by updating first and if it fails creating or vice versa. We cannot really know in advance since the data can always arrive at a destination via other sites in the ecosystem.